pkg/gardenlet/controller/networkpolicy/add.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package networkpolicy
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	predicateutils "github.com/gardener/gardener/pkg/controllerutils/predicate"
+	"github.com/gardener/gardener/pkg/gardenlet/controller/networkpolicy/helper"
+	"github.com/gardener/gardener/pkg/gardenlet/controller/networkpolicy/hostnameresolver"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "networkpolicy"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager, seedCluster cluster.Cluster) error {
+	if r.SeedClient == nil {
+		r.SeedClient = seedCluster.GetClient()
+	}
+	if r.shootNamespaceSelector == nil {
+		r.shootNamespaceSelector = labels.SelectorFromSet(labels.Set{
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleShoot,
+		})
+	}
+	if r.Resolver == nil {
+		resolver, err := hostnameresolver.CreateForCluster(seedCluster.GetConfig(), mgr.GetLogger())
+		if err != nil {
+			return fmt.Errorf("failed to get hostnameresolver: %w", err)
+		}
+		r.Resolver = resolver
+	}
+	if r.GardenNamespace == nil {
+		r.GardenNamespace = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: v1beta1constants.GardenNamespace,
+			},
+		}
+	}
+	if r.IstioSystemNamespace == nil {
+		r.IstioSystemNamespace = &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: v1beta1constants.IstioSystemNamespace,
+			},
+		}
+	}
```

Comment:
In other controllers, those reconciler attributes typically only contain the namespace name (`string`) and not the entire `*corev1.Namespace` object. Should we make this consistent?

Reply:
The parent `AddToManager` function of `gardenlet` contains the [entire namespace](https://github.com/gardener/gardener/blob/6c60ccc4f4c27ae767af1742e9581291473d5039/pkg/gardenlet/controller/add.go#L46) and so does [controllerinstallation](https://github.com/gardener/gardener/blob/6c60ccc4f4c27ae767af1742e9581291473d5039/pkg/gardenlet/controller/add.go#L72) controller. That's why I used it 😄

Reply:
The UID of garden namespace is needed here, https://github.com/gardener/gardener/blob/9fae12b4cc29928cefd5f0cade44290e33378844/pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go#L188, so the whole namespace is passed, which is not the case above.